pkg/dom0-ztools/efi/main.c

```diff
@@ -0,0 +1,697 @@
+      &gEfiCustomNextBootIndexGuid,
+      EFI_VARIABLE_NON_VOLATILE | EFI_VARIABLE_BOOTSERVICE_ACCESS,
+      Size,
+      &LastTriedBootOption);
+
+  return Status;
+}
+
+EFI_STATUS
+TryNextBoot(
+    VOID)
+{
+  EFI_STATUS Status;
+  UINT16 LastTriedBootOption, NextBoot = 0;
+  BOOLEAN IsValid = FALSE;
+
+  Status = GetLastTriedBootOption(&LastTriedBootOption);
+  if (EFI_ERROR(Status))
+  {
+    if (Status != EFI_NOT_FOUND)
+    {
+      return Status;
+    }
+    else
+    {
+      //
+      // Create and initialize the LastTriedBootIndex variable.
+      //
+      Status = SetLastTriedBootOption(0);
+      if (EFI_ERROR(Status))
+        return Status;
+    }
+  }
+
+  for (UINTN i = LastTriedBootOption + 1; i < gBootOrderSize / sizeof(UINT16); i++)
+  {
+    Status = IsValidBootOption(gBootOrder[i], &IsValid);
+    if (EFI_ERROR(Status))
+      return Status;
+
+    if (IsValid)
+    {
+      NextBoot = gBootOrder[i];
+      LastTriedBootOption = i;
+      break;
+    }
+  }
+
+  //
+  // No valid boot option left to try.
+  //
+  if (!IsValid)
+    return EFI_NOT_FOUND;
+
+  Status = SetNextBootOption(NextBoot);
+  if (EFI_ERROR(Status))
+    return Status;
+
+  // TOOD: Debug remove
+  DbgPrint(D_ERROR, (CHAR8 *)"===== Booting to : %d\n", NextBoot);
+
+  Status = SetLastTriedBootOption(LastTriedBootOption);
+  return Status;
+}
+
+
+//
+// Calling smbios->GetNext() results in either crash or hang,
+// so this one is more or less copy-pasted from EDKII's GetNext.
+//
+EFI_STATUS
+SmbiosGetNext(
+    CONST EFI_SMBIOS_PROTOCOL *This,
+    OUT EFI_SMBIOS_HANDLE *SmbiosHandle,
+    EFI_SMBIOS_TABLE_HEADER **Record)
+{
+  BOOLEAN StartPointFound;
+  LIST_ENTRY *Link;
+  LIST_ENTRY *Head;
+  SMBIOS_INSTANCE *Private;
+  EFI_SMBIOS_ENTRY *SmbiosEntry;
+  EFI_SMBIOS_TABLE_HEADER *SmbiosTableHeader;
+
+  if (SmbiosHandle == NULL)
+    return EFI_INVALID_PARAMETER;
+
+  StartPointFound = FALSE;
+  Private = SMBIOS_INSTANCE_FROM_THIS(This);
+  Head = &Private->DataListHead;
+  for (Link = Head->Flink; Link != Head; Link = Link->Flink)
+  {
+    SmbiosEntry = SMBIOS_ENTRY_FROM_LINK(Link);
+    SmbiosTableHeader = (EFI_SMBIOS_TABLE_HEADER *)(SmbiosEntry->RecordHeader + 1);
+
+    //
+    // If SmbiosHandle is 0xFFFE, the first matched SMBIOS record handle will be returned
+    //
+    if (*SmbiosHandle == SMBIOS_HANDLE_PI_RESERVED)
+    {
+      *SmbiosHandle = SmbiosTableHeader->Handle;
+      *Record = SmbiosTableHeader;
+      return EFI_SUCCESS;
+    }
+
+    //
+    // Start this round search from the next SMBIOS handle
+    //
+    if (!StartPointFound && (*SmbiosHandle == SmbiosTableHeader->Handle))
+    {
+      StartPointFound = TRUE;
+      continue;
+    }
+
+    if (StartPointFound)
+    {
+      *SmbiosHandle = SmbiosTableHeader->Handle;
+      *Record = SmbiosTableHeader;
+      return EFI_SUCCESS;
+    }
+  }
+
+  *SmbiosHandle = SMBIOS_HANDLE_PI_RESERVED;
+  return EFI_NOT_FOUND;
+}
+
+CHAR8 *
+GetSmbiosString(
```

Comment:
@rucoder (1/x) here I get the OEM string aka `type=11,value=BlahBlah`, aka `BlahBlah` in the smbios string table.